src/main/java/com/aws/greengrass/ggq/Exec.java

```diff
@@ -0,0 +1,466 @@
+                // path variable, but without using Exec shorthands to avoid initialization
+                // order paradoxes.
+                Process hack =
+                        Runtime.getRuntime().exec(new String[]{"sh", "-c", "echo 'echo $PATH' | grep -E ':[^ ]'"});
+                StringBuilder path = new StringBuilder();
+
+                Thread bg = new Thread(() -> {
+                    try (InputStream in = hack.getInputStream()) {
+                        for (int c = in.read(); c >= 0; c = in.read()) {
+                            path.append((char) c);
+                        }
+                    } catch (Throwable ignore) {
+                    }
+                });
+                bg.start();
+                bg.join(2000);
+                addPathEntries(path.toString().trim());
+                // Ensure some level of sanity
+                ensurePresent("/bin", "/usr/bin", "/sbin", "/usr/sbin");
+            }
+        } catch (Throwable ex) {
+            error("Error while initializing PATH", ex);
+        }
+        computePathString();
+    }
+    private static void error(String e, Throwable ex) {
+        System.out.println(e);
+        ex.printStackTrace(System.out);
+    }
+
+    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    Process process;
+    IntConsumer whenDone;
+    Consumer<CharSequence> stdout = NOP;
+    Consumer<CharSequence> stderr = NOP;
+    AtomicInteger numberOfCopiers;
+    private String[] environment = defaultEnvironment;
+    private String[] cmds;
+
+    private File dir = userdir;
+    private long timeout = -1;
+    private TimeUnit timeunit = TimeUnit.SECONDS;
+    private Copier stderrc;
+    private Copier stdoutc;
+
+    public static void setDefaultEnv(String key, CharSequence value) {
+        defaultEnvironment = setenv(defaultEnvironment, key, value, false);
+    }
+
+    private static String[] setenv(String[] env, String key, CharSequence value, boolean forceCopy) {
+        int elen = env.length;
+        int klen = key.length();
+        for (int i = 0; i < elen; i++) {
+            String s = env[i];
+            if (s.length() > klen && s.charAt(klen) == '=' && s.startsWith(key)) {
+                if (forceCopy) {
+                    env = Arrays.copyOf(env, env.length);
+                }
+                env[i] = key + '=' + value;
+                return env;
+            }
+        }
+        String[] ne = Arrays.copyOf(env, elen + 1, String[].class);
+        ne[elen] = key + '=' + value;
+        return ne;
+    }
+
+    public Exec setenv(String key, CharSequence value) {
+        environment = setenv(environment, key, value, environment == defaultEnvironment);
+        return this;
+    }
+
+    public static String cmd(String... command) throws InterruptedException, IOException {
+        return new Exec().withExec(command).execAndGetStringOutput();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *Exec* constructor in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).